gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/ConfigurableCleanableDataset.java

```diff
@@ -48,39 +50,58 @@
+
+    this.versionFinder =
+        ClassUtils.initClass(config.getString(VERSION_FINDER_CLASS_KEY), ImmutableList.of(FileSystem.class, Config.class),
+            ImmutableList.of(this.fs, config));
+    this.retentionPolicy =
+        ClassUtils.initClass(config.getString(RETENTION_POLICY_CLASS_KEY), ImmutableList.<Class<?>> of(Config.class),
+            ImmutableList.<Object> of(config));
+
+  }
+
   public ConfigurableCleanableDataset(FileSystem fs, Properties props, Path datasetRoot) throws IOException {
     this(fs, props, datasetRoot, LoggerFactory.getLogger(ConfigurableCleanableDataset.class));
   }
 
-  @SuppressWarnings("unchecked")
-  public ConfigurableCleanableDataset(FileSystem fs, Properties props, Path datasetRoot, Logger log)
-      throws IOException {
+  public ConfigurableCleanableDataset(FileSystem fs, Properties props, Path datasetRoot, Logger log) throws IOException {
     super(fs, props, log);
     this.datasetRoot = datasetRoot;
 
-    Preconditions.checkArgument(props.containsKey(VERSION_FINDER_CLASS_KEY),
-        "Missing property " + VERSION_FINDER_CLASS_KEY);
-    Preconditions.checkArgument(props.containsKey(RETENTION_POLICY_CLASS_KEY),
-        "Missing property " + RETENTION_POLICY_CLASS_KEY);
-
-    try {
-      Class<?> versionFinderClass = Class.forName(props.getProperty(VERSION_FINDER_CLASS_KEY));
-      Class<?> retentionPolicyClass = Class.forName(props.getProperty(RETENTION_POLICY_CLASS_KEY));
-      this.versionFinder = (VersionFinder) versionFinderClass.
-          getConstructor(FileSystem.class, Properties.class).newInstance(this.fs, props);
-      this.retentionPolicy = (RetentionPolicy) retentionPolicyClass.
-          getConstructor(Properties.class).newInstance(props);
-    } catch(ClassNotFoundException exception) {
-      throw new IOException(exception);
-    } catch(NoSuchMethodException exception) {
-      throw new IOException(exception);
-    } catch(InstantiationException exception) {
-      throw new IOException(exception);
-    } catch(IllegalAccessException exception) {
-      throw new IOException(exception);
-    } catch(InvocationTargetException exception) {
-      throw new IOException(exception);
-    }
+    Preconditions.checkArgument(props.containsKey(VERSION_FINDER_CLASS_KEY), "Missing property "
+        + VERSION_FINDER_CLASS_KEY);
+    Preconditions.checkArgument(props.containsKey(RETENTION_POLICY_CLASS_KEY), "Missing property "
+        + RETENTION_POLICY_CLASS_KEY);
+
+    this.versionFinder =
+        ClassUtils.initClass(props.getProperty(VERSION_FINDER_CLASS_KEY), ImmutableList.<Class<?>> of(Properties.class),
+            ImmutableList.<Object> of(props));
+    this.retentionPolicy =
+        ClassUtils.initClass(props.getProperty(RETENTION_POLICY_CLASS_KEY), ImmutableList.<Class<?>> of(Properties.class),
+            ImmutableList.<Object> of(props));
+  }
+
+  private ConfigurableCleanableDataset(FileSystem fs, Properties props, Path datasetRoot,
```

Comment:
The constructors can be used to setup versionFinder and RetentionPolicy in two ways.
1. Through properties
2. Through Config